crates/jstz_core/src/kv/transaction.rs

```diff
@@ -415,6 +375,141 @@ where {
     }
 }
 
+impl Transaction {
+    fn acquire_guard(&self) -> Result<RcGuardInner> {
+        if let Some(rc) = self.guard.borrow().upgrade() {
+            return Ok(rc);
+        }
+
+        let guard = self.inner.try_lock_arc().ok_or(KvError::LockPoisoned)?;
+        let rc = Rc::new(RefCell::new(guard));
+        *self.guard.borrow_mut() = Rc::downgrade(&rc);
+        Ok(rc)
+    }
+
+    /// Begin a transaction.
+    pub fn begin(&self) {
+        let rc = self.acquire_guard().expect("mutex poisoned");
+        rc.borrow_mut().begin();
+    }
+
+    /// Commit a transaction.
+    pub fn commit(&self, rt: &mut impl Runtime) -> Result<()> {
+        let rc = self.acquire_guard()?;
+        let mut inner = rc.borrow_mut();
+
+        inner.commit(rt)
+    }
+
+    /// Rollback a transaction.
+    pub fn rollback(&self) -> Result<()> {
+        let rc = self.acquire_guard()?;
+        let mut inner = rc.borrow_mut();
+
+        inner.rollback()
+    }
+
+    pub fn get<'a, V: Value>(
+        &'a self,
+        rt: &impl Runtime,
+        key: Key,
+    ) -> Result<Option<Guarded<'a, V>>> {
+        let rc = self.acquire_guard()?;
+        let mut inner = rc.borrow_mut();
+
+        match inner.lookup::<V>(rt, key)? {
+            Some(entry) => {
+                let value = entry.as_ref()?;
+                Ok(Some(Guarded::new(rc.clone(), value)))
+            }
+            None => Ok(None),
+        }
+    }
+
+    pub fn get_mut<'a, V: Value>(
+        &'a self,
+        rt: &impl Runtime,
+        key: Key,
+    ) -> Result<Option<GuardedMut<'a, V>>> {
+        let rc = self.acquire_guard()?;
+        let mut inner = rc.borrow_mut();
+
+        match inner.lookup_mut::<V>(rt, key)? {
+            Some(entry) => {
+                let value = entry.as_mut()?;
+                Ok(Some(GuardedMut::new(rc.clone(), value)))
+            }
+            None => Ok(None),
+        }
+    }
+
+    pub fn contains_key(&self, rt: &impl Runtime, key: &Key) -> Result<bool> {
+        let rc = self.acquire_guard()?;
+        let inner = rc.borrow_mut();
+        inner.contains_key(rt, key)
+    }
+
+    pub fn insert<V: Value>(&self, key: Key, value: V) -> Result<()> {
+        let rc = self.acquire_guard()?;
+        let mut inner = rc.borrow_mut();
+        inner.current_snapshot_insert(key, SnapshotValue::new(value))
+    }
+
+    pub fn remove(&self, key: Key) -> Result<()> {
+        let rc = self.acquire_guard()?;
+        let mut inner = rc.borrow_mut();
+        inner.current_snapshot_remove(key)
+    }
+
+    /// Returns the given key's corresponding entry in the transactional
+    /// snapshot for in-place manipulation.
+    pub fn entry<'a, 'b, V>(
+        &'a mut self,
+        rt: &impl Runtime,
+        key: Key,
+    ) -> Result<Entry<'b, V>>
+    where
+        V: Value,
+        'a: 'b,
```

Comment:
Its morally incorrect to decouple the lifetime of the `Transaction` and `Entry`
```suggestion
    pub fn entry<'a, V>(
        &'a mut self,
        rt: &impl Runtime,
        key: Key,
    ) -> Result<Entry<'a, V>>
    where
        V: Value
```